Allow implicit type conversion in `Array` to `QBit` cast

rienath opened the pull request:
Previously, casting `Array` to `QBit` required the array element type to exactly match the `QBit` element type. For example:

```sql
CREATE TABLE test (qbit QBit(Float32, 3)) ENGINE = Memory;
INSERT INTO test SELECT [1, 2, 3];  -- Failed with TYPE_MISMATCH
```

Now array elements are implicitly converted to the target `QBit` element type using the standard type conversion infrastructure, consistent with how other type conversions work in ClickHouse. Here is an example of a few more queries that now work:

```sql
SELECT [1, 2, 3]::QBit(Float64, 3);                       -- Array(UInt8) → QBit(Float64, 3)
SELECT CAST([0.1, 1.1] AS QBit(BFloat16, 2));             -- Array(Float64) → QBit(BFloat16, 2)
```

and even

```sql
INSERT INTO test SELECT arrayMap(i -> i + number, range(3)) AS qbit FROM numbers(3);
```

Closes: https://github.com/ClickHouse/ClickHouse/issues/89663.

### Changelog category (leave one):
- Improvement


### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Allow implicit type conversion when casting `Array` to `QBit`. Integer and float arrays can now be inserted directly into `QBit` columns without explicit type casts.

rienath committed:
Allow implicit type conversion for Array to QBit cast

rienath committed:
Simplify QBit tests using implicit array element conversion

Remove explicit type casts (toFloat32, toFloat64, toBFloat16) in test queries now that Array to QBit conversion handles implicit element type conversion automatically.

rienath commented on tests/queries/0_stateless/03365_qbit_casts_as_from_array.sql:
```diff
@@ -9,17 +9,17 @@ SELECT * FROM format('Values', 'qbit QBit(Float64, 2)', '(array(0.1,1.1))');
-SELECT CAST((SELECT groupArray(toFloat32(number + 0.1)) FROM numbers(2)) AS QBit(Float32, 2));
+SELECT CAST((SELECT groupArray(number + 0.1) FROM numbers(2)) AS QBit(Float32, 2));
+SELECT CAST((SELECT groupArray(number + 0.1) FROM numbers(2)) AS QBit(Float32, 2));
+SELECT CAST((SELECT groupArray(number + 0.1) FROM numbers(2)) AS QBit(Float32, 2));
```
Simplified a few that have been using implicit conversion

rienath committed:
Fix dangling reference in lambda

rienath committed:
Preserve nullable

rienath committed:
Ignore clang-tidy

rienath committed:
Minor test fix

pamarcos commented on src/Functions/FunctionsConversion.h:
```diff
@@ -4892,26 +4892,28 @@ class FunctionCast final : public IFunctionBase
+        const size_t dimension = to_qbit_type.getDimension();
+        const size_t element_size = to_qbit_type.getElementSize();
+
+        return [this, from_nested_type, to_nested_type, dimension, element_size]
```
This is in a few other methods as well

rienath commented on src/Functions/FunctionsConversion.h:
```diff
@@ -4927,26 +4892,28 @@ class FunctionCast final : public IFunctionBase
+             const ColumnNullable * nullable_source,
+             size_t /* input_rows_count */) -> ColumnPtr
+        {
+            const auto nested_function = prepareUnpackDictionaries(from_nested_type, to_nested_type);
```
To avoid larger capture. But the difference is marginal and we could avoid computing it multiple times by having it outside, I will move it

rienath committed:
Optimise createArrayToQBitWrapper

(1) Remove explicit copying of from/to_nested_type before lambda to avoid 'NOLINT' as we copy in lamba capture anyway, (2) move 'prepareUnpackDictionaries(..)' and creation of to_array_type outside lambda (larger capture, but computation)

rienath committed:
Refactor performance tests

We no longer need toBFloat16/toFloat32/toFloat64 functions for QBit to operate

rienath committed:
Merge remote-tracking branch 'origin' into qbit-usability

rienath reviewed (commented):
@pamarcos thanks for the review! I wasn't replying until I see that the CI is green to save your time :)

rienath commented:
We usually don't backport improvements, but this one makes user experience much nicer and Christmas / New Year is soon. Let's make it a gift 🎁

clickgapai commented:
Hi @rienath @pamarcos — while reviewing this PR I found the following:

- **Bug (P1):** `accurateCastOrNull` Array→QBit returns `[0]` instead of NULL when element conversion is inaccurate — https://github.com/ClickHouse/ClickHouse/issues/103078

Happy to discuss — close anything that's wrong or already addressed.
